Code/Source/Frame/ROS2Transform.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+* Copyright (c) Contributors to the Open 3D Engine Project.
+* For complete copyright and license terms please see the LICENSE at the root of this distribution.
+*
+* SPDX-License-Identifier: Apache-2.0 OR MIT
+*
+*/
+
+#include "Frame/ROS2Transform.h"
+#include "Utilities/ROS2Conversions.h"
+#include "ROS2/ROS2Bus.h"
+
+namespace ROS2
+{
+    namespace Internal
+    {
+        class StaticTransformPublisher : public TransformPublisher
+        {
+        public:
+            void Publish(const geometry_msgs::msg::TransformStamped& transformMessage) override
+            {
+                if (m_published)
+                    return; // Consider adding a warning here as well
+
+                return ROS2Interface::Get()->BroadcastStaticTransform(transformMessage);
+                m_published = true;
+            }
+
+        private:
+            bool m_published = false;
+        };
+
+        class DynamicTransformPublisher : public TransformPublisher
+        {
+        public:
+            DynamicTransformPublisher()
+            {
+                auto ros2Node = ROS2Interface::Get()->GetNode();
+
+                // TODO - determine whether default QoS is good and what is good practice for queue size. Compare to tf2 broadcaster
+                const size_t keepLastQueueSize = 10;
+                m_dynamicTFPublisher = ros2Node->create_publisher<geometry_msgs::msg::TransformStamped>("/tf", keepLastQueueSize);
+            }
+
+            void Publish(const geometry_msgs::msg::TransformStamped& transformMessage) override
+            {
+                m_dynamicTFPublisher->publish(transformMessage);
+            }
+
+        private:
+            rclcpp::Publisher<geometry_msgs::msg::TransformStamped>::SharedPtr m_dynamicTFPublisher;
+        };
+
+        AZStd::unique_ptr<TransformPublisher> TransformPublisher::CreateTransformPublisher(bool isDynamic)
+        {
+            if (isDynamic)
+            {
+                return std::make_unique<DynamicTransformPublisher>();
+            }
+            else
+            {
+                return std::make_unique<StaticTransformPublisher>();
+            }
+        }
+    }
+
+    ROS2Transform::ROS2Transform(const AZStd::string& parentFrame, const AZStd::string& childFrame, bool isDynamic)
+        : m_parentFrame(parentFrame), m_childFrame(childFrame)
```

Comment:
optimization:" You could accept parentFrame and childFrame by value and then move them into the `m_parentFrame` and `m_childFrame` member variables.

Reply:
Could you explain reasoning behind this suggestion?
I see two possible advantages for your suggestion here:
1) would make no copy for rvalue arguments (2 moves?)
2) would make it explicit that function makes a copy of frame arguments (in signature, without needing to look elsewhere).

Anyway, I would like to understand your reasoning since I am not used to optimizing functions that are called this rarely.

Reply:
The reason for my suggestion is based on how this function could be called.

### Scenario 1 - const char(&)[N] string literal
If the user constructs `ROS2Transform` with string literal parameters for the `parentFrame` and `childFrame` than a `AZStd::string` temporary instance has to be created and abound to the `const AZStd::string&` type.
Afterwards the copy constructor of the AZStd::string class is then invoked to make a copy of the temporary instance.

```
ROS2Transform transform1("Parent", "Child", true);
// before the call - Two AZStd::strings are constructed and bound to the ROS2Transform constructor parameters 
// during the call - Two copy constructor calls are invoked to store the m_parentFrame and m_childFrame members
```

With the signature of `const AZStd::string` signature, the AZStd::string `const char*` constructor has to be invoked 2 times + the AZStd::string copy constructor has to be invoked 2 times as well
So total operations = 4 copies

Using a signature of `AZStd::string`, the AZStd::string `const char*` constructor has to be invoked 2 times + the AZStd::string move constructor has to be invoked 2 times as well.
So total operations = 2 copies + 2 moves

In this scenario the `AZStd::string` signature wins

### Scenario 1 - AZStd::string copy
Next let's move on to the situation where the constructor is called with an `AZStd::string` call where a copy is wanted.
```
AZStd::string memberParentFrame = "MemberParent";
AZStd::string memberChildFrame = "MemberChild";
// ...
ROS2Transform transform2(memberParentFrame, memberChildFrame, true);
```

With the signature of `const AZStd::string&`, the `memberParentFrame` and `memberChildFrame` variables are bound to reference, so that operation is free, but because the parameter is `const AZStd::string&`, the ROS2Transform constructor has to make a copy
So total operations = 2 copies

With the signature of `AZStd::string`, the `memberParentFrame` and `memberChildFrame` variables values would have to  invoke the copy constructor to bind it to the `parentFrame` and `childFrame` parameters, but inside of the ROS2Transform constructor only a move constructor operation has to be performed.
So total operations  = 2 copies + 2 moves

In this scenario `const AZStd::string&` signature wins

### Scenario 3 - AZStd::string move
In this situation there is an AZStd::string variable on the stack, but it is moved during the call to the ROS2Transform constructor
```
AZStd::string transientParentFrame = "TransientParent";
AZStd::string transientChildFrame = "TransientChild";
// ...
ROS2Transform transform3(AZStd::move(transientParentFrame), AZStd::move(transientChildFrame), true);
```

Now if the ROS2Transform constructor had the signature of  `const AZStd::string&`, the `AZStd::move(transientParentFrame)` and `AZStd::move(transientChildFrame)` expression would cast to the type of `AZStd::string&&` which due to reference binding rules is allowed to bind to `const AZStd::string&`.
Now because the parameter signature is still `const AZStd::string&`, the ROS2Transform cosntructor still has to make a copy.
So total operations = 2 copies

Now if the ROS2Transform constructor had the signature of `AZStd::string`, the `AZStd::move(transientParentFrame)` and `AZStd::move(transientChildFrame)` expressions would cause the move constructor of `AZStd::string` to be invoked when calling the constructor.
Afterwards inside of the constructor, the `parentFrame` and `childFrame` parameters can then be moved into the member variables. In this case there are 4 move operations
So total operation = 4 moves

In this scenario `AZStd::string` signature wins.
4 moves are better than 2 copies, since a move only has to copy the data pointers, size and capacity values, while a copy has to allocate a block of memory, copy over the string data + copy over the size and capacity values

Reply:
I can change this constructor